core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -0,0 +1,220 @@
+        if (Files.exists(keystoreFile) == false) {
+            return null;
+        }
+        DataInputStream inputStream = new DataInputStream(Files.newInputStream(keystoreFile));
+        int format = inputStream.readInt();
+        if (format != FORMAT_VERSION) {
+            throw new IllegalStateException("Unknown keystore metadata format [" + format + "]");
+        }
+        boolean hasPassword = inputStream.readBoolean();
+        String type = inputStream.readUTF();
+        String secretKeyAlgo = inputStream.readUTF();
+        return new KeyStoreWrapper(hasPassword, type, secretKeyAlgo, inputStream);
+    }
+
+    /** Returns true iff {@link #loadKeystore(char[])} has been called. */
+    public boolean isLoaded() {
+        return keystore.get() != null;
+    }
+
+    /** Return true iff calling {@link #loadKeystore(char[])} requires a non-empty password. */
+    public boolean hasPassword() {
+        return hasPassword;
+    }
+
+    /** Loads the keystore this metadata wraps. This may only be called once. */
+    public void loadKeystore(char[] password) throws GeneralSecurityException, IOException {
+        this.keystore.set(KeyStore.getInstance(type));
+        try (InputStream in = input) {
+            keystore.get().load(in, password);
+        }
+
+        this.keystorePassword.set(new KeyStore.PasswordProtection(password));
+        Arrays.fill(password, '\0');
+
+        // convert keystore aliases enum into a set for easy lookup
+        Enumeration<String> aliases = keystore.get().aliases();
+        while (aliases.hasMoreElements()) {
+            settingNames.add(aliases.nextElement());
+        }
+    }
+
+    /** Write the keystore to the given config directory. */
+    void save(Path configDir) throws Exception {
+        Path keystoreFile = keystorePath(configDir);
+        try (DataOutputStream outputStream = new DataOutputStream(Files.newOutputStream(keystoreFile))) {
+            outputStream.writeInt(FORMAT_VERSION);
+            char[] password = this.keystorePassword.get().getPassword();
+            outputStream.writeBoolean(password.length != 0);
+            outputStream.writeUTF(type);
+            outputStream.writeUTF(secretFactory.getAlgorithm());
+            keystore.get().store(outputStream, password);
+        }
+        PosixFileAttributeView attrs = Files.getFileAttributeView(keystoreFile, PosixFileAttributeView.class);
+        if (attrs != null) {
+            // don't rely on umask: ensure the keystore has minimal permissions
+            attrs.setPermissions(PosixFilePermissions.fromString("rw-------"));
+        }
+    }
+
+    /** Returns the names of all settings in this keystore. */
+    public Set<String> getSettings() {
+        return settingNames;
+    }
+
+    /** Retrieve a string setting. The {@link SecureString} should be closed once it is used. */
+    SecureString getStringSetting(String setting) throws GeneralSecurityException {
```

Comment:
I wonder if we should think about adding a special security permission that we have to grant to all plugins that want to access the keystore. That way we can prevent unintended access to it and users will be prompted when they install the plugin. Only certain plugins should be able to access this stuff right?